### ISSUE TITLE
Wrong declaration of GatsbyLink with typescript 2.4.2

### ISSUE THREAD

**Issue by DominikGuzei** [gatsbyjs/gatsby#1627]
Hey there, im trying to use Gatsby with typescript 2.4.2 but it throws an error telling me that the [signature of the `GatsbyLink` export](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-link/index.d.ts#L8)  is wrong:

```ts
export default class GatsbyLink extends React.Component<GatsbyLinkProps, void> { }
```

I could fix it by changing it to
```ts
export default class GatsbyLink extends React.Component<GatsbyLinkProps> { }
```

Im just running the starter code for `index.tsx`:
```tsx
import * as React from 'react';
import Link from 'gatsby-link';

const IndexPage = () => (
  <div>
    <h1>Hi people</h1>
    <p>Welcome to your new Gatsby site.</p>
    <p>Now go build something great.</p>
    <Link to="/page-2/">Go to page 2</Link>
  </div>
);

export default IndexPage;
```

The error message:
```
Error:(9, 5) TS2605:JSX element type 'GatsbyLink' is not a constructor function for JSX elements.
  Types of property 'setState' are incompatible.
    Type '{ <K extends never>(f: (prevState: void, props: GatsbyLinkProps) => Pick<void, K>, callback?: (()...' is not assignable to type '{ <K extends never>(f: (prevState: {}, props: any) => Pick<{}, K>, callback?: (() => any) | undef...'.
      Types of parameters 'f' and 'f' are incompatible.
        Types of parameters 'prevState' and 'prevState' are incompatible.
          Type 'void' is not assignable to type '{}'.
```

my tsconfig:
```json
{
  "compilerOptions": {
    "outDir": "./dist/",
    "sourceMap": true,
    "noImplicitAny": true,
    "strict": true,
    "strictNullChecks": true,
    "module": "commonjs",
    "target": "esnext",
    "jsx": "react",
    "lib": ["dom", "es2015", "es2017"]
  },
  "include": [
    "./src/**/*"
  ]
}
```

See last comment on [this issue](https://github.com/ant-design/ant-design/issues/4892) to see how i found the fix.

**Comment by KyleAMathews** [comment]
@timsuchanek and @fabien0102 comments?

**Comment by timsuchanek** [comment]
I think the problem is that Typescript got stricter about the generics, in this case the `void`. With removing the `void`, it's then using the default `{}`. (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L238)
Do you want to create a PR for that @DominikGuzei ?

**Comment by DominikGuzei** [comment]
@timsuchanek yeah sure, will set it up!

**Comment by DominikGuzei** [comment]
@timsuchanek the fix is ready for review: https://github.com/gatsbyjs/gatsby/pull/1628

**Comment by KyleAMathews** [comment]
Fixed in #1628
